### ISSUE TITLE
[a11y][Violation][2.4.1 Bypass Blocks] Element with "banner" role missing `aria-labelledby` attribute

### ISSUE THREAD

**Issue by nkn2022** [jupyterlab/jupyterlab#16869]
## Description

I am running an accessibility test suite with Jupyter using the **IBM Equal Access Accessibility Checker** tool on Chrome browser. The results look good overall, only a few things are missing to reach a first formal level of compliance. So kindly help look into these. 

Issues reported were violating **WCAG 2.2** requirements. 
Ref: https://www.ibm.com/able/requirements/checker-rule-sets

Listing of violations categorized based on the checkpoints. 
Full report (includes information on element locations where issue were found): 
[Accessibility_Report-Untitled.ipynb - JupyterLab.xlsx](https://github.com/user-attachments/files/17383160/Accessibility_Report-Untitled.ipynb.-.JupyterLab.xlsx)

# Category: `2.4.1 Bypass Blocks`
Ref: https://www.ibm.com/able/requirements/requirements/?version=v7_3#2_4_1

****Violation of Rule:** `aria_banner_label_unique`**
> List of issues with elements that violate this rule with a screenshots
> > **Issue 1** 
> > Multiple elements with "banner" role do not have unique ARIA labels
> **Element location:** 
> > `<div style="position: absolute; ... d-hover: #e0e0e0;" id="jp-top-panel" role="banner" class="lm-Widget lm-Panel">`
> > **Screenshot:**
> > <img width="468" alt="image" src="https://github.com/user-attachments/assets/35565c4c-99b4-40c8-b08f-0b70fe511ea2">
> > **PS:** You will not see this issue reported in the attached file. This is only reported in our product where the JupyterLab UI is within another container with a `<header / >` element. So we will need a label definition on this component to fix this problem. 

## Reproduce

- Install Chrome extension for IBM accessibility checker tool. Ref: https://www.ibm.com/able/toolkit/verify/automated
- Open JupyterLab UI on Chrome browser. Open `notebook` file to edit.
- Run the tool and we can see the above `Violation`s reported.

Here is the screenshot of the UI scanned for the attached report:

<img width="1534" alt="image" src="https://github.com/user-attachments/assets/0e0cbc5c-cf21-4393-97fe-b566da24b58b">

## Expected behavior
Here is the suggestion from the tool:

> **What to do**
> For each `<header>` element that is a descendant of the `<body>` element, and each element with role="banner", either set the value of the aria-labelledby attribute of each banner to the id value of any visible text that appropriately and uniquely describes that banner's purpose
> Or, use the aria-label attribute to add an invisible, unique, non-empty label to each element with the "banner" role
> For example:
> 
> `<h1 id="weather">Weather portlet</h1>`
> `<header aria-labelledby="weather">`
>    ...
> `</header>`
>     
> `<div role="banner" aria-label="Stock ticker portlet">`
>    ...
> `</div>`

## Context

- Operating System and version: MacOS Sonoma Version 14.5
- Browser and version: Chrome Version 129+
- JupyterLab version: Version 4.2.5

**Comment by nkn2022** [comment]
@JasonWeill Thanks for checking. Happy to help. I will open separate ones with screenshots soon.

**Comment by nkn2022** [comment]
I have created one issue for each category of the issues with details and screenshots. Kindly let me know if this breakdown looks good.

I am changing the description and title of this issue to track only issues found in category `2.4.1 Bypass Blocks`.

Created the following to track the remaining issues:
`1.4.3 Contrast (Minimum)` - https://github.com/jupyterlab/jupyterlab/issues/16888
`2.5.8 Minimum Target Size` - https://github.com/jupyterlab/jupyterlab/issues/16889
`4.1.2 Name, Role, Value` - https://github.com/jupyterlab/jupyterlab/issues/16890
`ARIA specification` - https://github.com/jupyterlab/jupyterlab/issues/16891

**Comment by JasonWeill** [comment]
Thank you very much! I'm going to close this issue in favor of the four other ones that you've opened. We do weekly issue triage at 09:00 US Pacific ([in your time zone](https://www.timeanddate.com/worldclock/fixedtime.html?msg=Triage&iso=20241029T09&p1=234)) on Tuesdays, if you'd like to join us next week. I appreciate your contributions!

**Comment by JasonWeill** [comment]
Sorry about that — I closed this in error, but noticed that it was retitled to cover one issue. Reopened so that we can revisit it at triage next week.

**Comment by nkn2022** [comment]
@JasonWeill Thanks! Yes, I kept this is as one of the issue.

**Comment by JasonWeill** [comment]
Thank you @nkn2022! I now see three sub-issues:

1. Multiple elements with "banner" role do not have unique ARIA labels — This is clear and actionable
2. Content is not within a landmark element — This is not clear, because the "Skip to main section" container is not necessarily a landmark element.
3. Multiple elements with "toolbar" roles do not have unique labels — How should we label our toolbars?

It might be easier to complete this work if we further break this issue down into the three sub-issues listed above. Sub-issue 1 is something that's ready to work on. @nkn2022 if you could please file separate issues for sub-issue 2 and 3 above?

**Comment by JasonWeill** [comment]
Sub-issue 1 (adding ARIA labels to "banner" elements) was originally addressed in #9622, but more work appears to be needed.

**Comment by nkn2022** [comment]
@JasonWeill The issue #2 appears only if the JupyterUI is scanned independently. 
See:
<img width="1531" alt="image" src="https://github.com/user-attachments/assets/18d0f6b7-eb0b-4766-bf8b-26b00ef67fd7">

Does not seem to be reported as a Violation in our embedded UI. So I will not need to track this issue.  So will not open a new one for that.

**Comment by nkn2022** [comment]
Opened separate issue for the toolbar problem: https://github.com/jupyterlab/jupyterlab/issues/16905

**Comment by Rishab87** [comment]
I've created a PR for it! Can anyone please review it?

**Comment by krassowski** [comment]
Hi @nkn2022 this has not been released yet, we have only three commits on 4.3.x branch since the last release (including this one) so I would like to wait a little bit longer before cutting another one, but it will go into 4.3.6, likely to be out later this month.

**Comment by nkn2022** [comment]
@krassowski Is it okay to keep this issue open and update when a release is available? To avoid confusion.

**Comment by krassowski** [comment]
@nkn2022 the usual workflow on GitHub is to close issues as PRs are merged. 4.3.6 is now available: https://github.com/jupyterlab/jupyterlab/releases/tag/v4.3.6

**Comment by nkn2022** [comment]
@krassowski Great! Thank you for the update.

**Comment by nkn2022** [comment]
Verified on 4.3.6. This issue is not reported any more. Thanks!
